sha512_mb/sha512_ctx_base.c

```diff
@@ -150,41 +148,87 @@ static void sha512_init(SHA512_HASH_CTX * ctx, const void *buffer, uint32_t len)
 	ctx->status = HASH_CTX_STS_PROCESSING;
 }
 
-static uint32_t sha512_update(SHA512_HASH_CTX * ctx, const void *buffer, uint32_t len)
+static uint32_t sha512_update(SHA512_HASH_CTX *ctx, const void *buffer, uint32_t len)
 {
 	uint32_t remain_len = len;
 	uint64_t *digest = ctx->job.result_digest;
 
-	while (remain_len >= SHA512_BLOCK_SIZE) {
-		sha512_single(buffer, digest);
-		buffer = (void *)((uint8_t *) buffer + SHA512_BLOCK_SIZE);
-		remain_len -= SHA512_BLOCK_SIZE;
-		ctx->total_length += SHA512_BLOCK_SIZE;
+	// Store buffer ptr info from user
+	ctx->incoming_buffer = buffer;
+	ctx->incoming_buffer_length = len;
+
+	// Advance byte counter
+	ctx->total_length += len;
+
+	while (remain_len) {
+		// If there is anything currently buffered in the extra blocks, append to it until it contains a whole block.
+		// Or if the user's buffer contains less than a whole block, append as much as possible to the extra block.
+		if ((ctx->partial_block_buffer_length) | (remain_len < SHA512_BLOCK_SIZE)) {
+			// Compute how many bytes to copy from user buffer into extra block
+			uint32_t copy_len = SHA512_BLOCK_SIZE - ctx->partial_block_buffer_length;
+			if (remain_len < copy_len) {
+				copy_len = remain_len;
+			}
+
+			if (copy_len) {
+				const void *pbuf = ctx->incoming_buffer;
+				// Copy and update relevant pointers and counters
+				memcpy_fixedlen(&ctx->partial_block_buffer
+						[ctx->partial_block_buffer_length], pbuf, copy_len);
+
+				ctx->partial_block_buffer_length += copy_len;
+				remain_len -= copy_len;
+				ctx->incoming_buffer = (const void *)((const char *)pbuf + copy_len);
+				ctx->incoming_buffer_length = remain_len;
+			}
+			// The extra block should never contain more than 1 block here
+			assert(ctx->partial_block_buffer_length <= SHA512_BLOCK_SIZE);
+
+			// If the extra block buffer contains exactly 1 block, it can be hashed.
+			if (ctx->partial_block_buffer_length >= SHA512_BLOCK_SIZE) {
+				ctx->partial_block_buffer_length = 0;
+				sha512_single(ctx->partial_block_buffer, digest);
+			}
+		}
+
+		// If the extra blocks are empty, begin hashing what remains in the user's buffer.
+		if (ctx->partial_block_buffer_length == 0 && ctx->incoming_buffer_length) {
+			uint8_t *pbuf = (uint8_t *)(ctx->incoming_buffer);
+			uint32_t blocks = ctx->incoming_buffer_length >> SHA512_LOG2_BLOCK_SIZE;
+
+			while (blocks--) {
+				sha512_single((const void *)pbuf, digest);
+				pbuf += SHA512_BLOCK_SIZE;
+				remain_len -= SHA512_BLOCK_SIZE;
+			}
+			ctx->incoming_buffer = (const void *)pbuf;
+			ctx->incoming_buffer_length = remain_len;
+		}
```

Comment:
Could you refactor this part? I think below structure might be better.
```
/* process partial block */
....
/* If the extra blocks are empty, begin hashing what remains in the user's buffer. */
...
/* process remain_len < block_size */
....
```

Reply:
ok

Reply:
My suggestion:
```

static void sha512_update(SHA512_HASH_CTX *ctx, const void *buffer, uint32_t len)
{
	uint32_t remain_len = len;
	uint64_t *digest = ctx->job.result_digest;

	// Advance byte counter
	ctx->total_length += len;

	// If there is anything currently buffered in the extra blocks, append to it until it contains a whole block.
	// Or if the user's buffer contains less than a whole block, append as much as possible to the extra block.
	if ((ctx->partial_block_buffer_length) | (remain_len < SHA512_BLOCK_SIZE)) {
		// Compute how many bytes to copy from user buffer into extra block
		uint32_t copy_len = SHA512_BLOCK_SIZE - ctx->partial_block_buffer_length;
		if (remain_len < copy_len) {
			copy_len = remain_len;
		}

		if (copy_len) {
			const void *pbuf = buffer;
			// Copy and update relevant pointers and counters
			memcpy_fixedlen(&ctx->partial_block_buffer
					[ctx->partial_block_buffer_length], pbuf, copy_len);

			ctx->partial_block_buffer_length += copy_len;
			remain_len -= copy_len;
			buffer = (void *)((uint8_t *) buffer + copy_len);
		}
		// The extra block should never contain more than 1 block here
		assert(ctx->partial_block_buffer_length <= SHA512_BLOCK_SIZE);

		// If the extra block buffer contains exactly 1 block, it can be hashed.
		if (ctx->partial_block_buffer_length >= SHA512_BLOCK_SIZE) {
			ctx->partial_block_buffer_length = 0;
			sha512_single(ctx->partial_block_buffer, digest);
		}
	}

	// If the extra blocks are empty, begin hashing what remains in the user's buffer.
	if ( ctx->partial_block_buffer_length == 0 )
	{
		while ( remain_len >= SHA512_BLOCK_SIZE ) {
			sha512_single(buffer, digest);
			buffer = (void *)((uint8_t *) buffer + SHA512_BLOCK_SIZE);
			remain_len -= SHA512_BLOCK_SIZE;
		}

	}

	if ( remain_len > 0 )
	{
		memcpy_fixedlen(&ctx->partial_block_buffer, buffer, remain_len);
		ctx->partial_block_buffer_length = remain_len;
	}

	ctx->status = HASH_CTX_STS_IDLE;
	return ;
}
```

Reply:
yes, it's more clear. thanks.